Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,177 @@
+
+(defclass redefinition-updates-instances-class (standard-class)
+  ((%instances :accessor instances
+               :initform '())))
+
+(defmethod c2mop:validate-superclass
+    ((class      redefinition-updates-instances-class)
+     (superclass standard-class))
+  t)
+
+(defmethod shared-initialize :before
+    ((instance   redefinition-updates-instances-class)
+     (slot-names t)
+     &key (direct-superclasses nil direct-superclasses-supplied-p))
+  (when direct-superclasses-supplied-p
+    (unless (some (alexandria:rcurry #'subtypep 'application-frame)
+                  direct-superclasses)
+      (error "~@<Instances of ~S must be subclasses of ~S.~@:>"
+             'redefinition-updates-instances-class 'application-frame))))
+
+(defmethod reinitialize-instance :after
+    ((instance redefinition-updates-instances-class) &key)
+  ;; Trigger redefinition notification protocol which will directly or
+  ;; call `handle-frame-class-redefinition' or schedule it for later.
+  (note-frame-class-redefined instance))
+
+(defmethod make-instance :around ((class redefinition-updates-instances-class)
+                                  &key)
+  ;; Store a weak pointer to the created instance for updating it when
+  ;; CLASS is redefines.
+  (let ((instance (call-next-method)))
+    (push (tg:make-weak-pointer instance) (instances class))
+    instance))
+
+(defmethod handle-frame-class-redefinition
+    ((class redefinition-updates-instances-class))
+  (let ((remaining         '())
+        (new-layouts       nil)
+        (new-command-table nil))
+    (labels ((initarg-value (name)
+               (alexandria:when-let* ((initargs (c2mop:class-default-initargs class))
+                                      (initarg  (find name initargs :key #'first)))
+                 (funcall (third initarg))))
+             (new-layouts ()
+               (or new-layouts
+                   (setf new-layouts (initarg-value :layouts))))
+             (new-command-table ()
+               (or new-command-table
+                   (setf new-command-table (initarg-value :command-table))))
+             (update (instance)
+               (with-simple-restart (continue "~@<Skip updating frame ~A~@:>"
+                                              instance)
+                 (event-queue-append
+                  (frame-event-queue instance)
+                  (make-instance 'frame-class-redefined-event
+                                 :sheet             instance
+                                 :new-layouts       (new-layouts)
+                                 :new-command-table (new-command-table)))))
+             (maybe-update (weak-instance)
+               ;; Update instances that haven't become garbage and
+               ;; haven't been CHANGE-CLASSed. Implicitly drop other
+               ;; instances.
+               (when-let ((live-instance (tg:weak-pointer-value weak-instance)))
+                 (when (typep live-instance class)
```

Comment:
are we expecting frames to have changed class?

Reply:
Yes, a user could do that.